File: patterns/01-design-patterns/03-behavioral/08-iterator/python/iterator.py
class PatternError(Exception):
    def __init__(self,code):super().__init__(code);self.code=code
class PageIterator:
    def __init__(self,pages):self.pages=pages;self.page_index=0;self.items=[];self.item_index=0;self.fetched=0
    def __next__(self):
        while self.item_index>=len(self.items):
            if self.page_index>=len(self.pages):raise StopIteration
            page=self.pages[self.page_index];self.page_index+=1
            if isinstance(page,dict):raise PatternError(page["error"])
            self.items=list(page);self.item_index=0;self.fetched+=1
        value=self.items[self.item_index];self.item_index+=1;return value
def evaluate(input_data):
    iterator=PageIterator(input_data.get("pages",[]));items=[];exhausted=False
    for _ in range(input_data.get("take",0)):
        try:items.append(next(iterator))
        except StopIteration:exhausted=True;break
    return{"items":items,"fetched_pages":iterator.fetched,"exhausted":exhausted}

File: patterns/01-design-patterns/03-behavioral/08-iterator/go/iterator.go
package iterator

type PatternError struct{ code string }

func (e *PatternError) Error() string { return e.code }
func (e *PatternError) Code() string  { return e.code }

type PageIterator struct {
	pages                         []any
	pageIndex, itemIndex, fetched int
	items                         []any
}

func (i *PageIterator) Next() (int, bool, error) {
	for i.itemIndex >= len(i.items) {
		if i.pageIndex >= len(i.pages) {
			return 0, false, nil
		}
		page := i.pages[i.pageIndex]
		i.pageIndex++
		if failure, ok := page.(map[string]any); ok {
			return 0, false, &PatternError{code: failure["error"].(string)}
		}
		i.items = page.([]any)
		i.itemIndex = 0
		i.fetched++
	}
	value := int(i.items[i.itemIndex].(float64))
	i.itemIndex++
	return value, true, nil
}
func Evaluate(input map[string]any) (any, error) {
	pages, _ := input["pages"].([]any)
	iterator := &PageIterator{pages: pages}
	items := []int{}
	exhausted := false
	take := int(input["take"].(float64))
	for n := 0; n < take; n++ {
		value, ok, err := iterator.Next()
		if err != nil {
			return nil, err
		}
		if !ok {
			exhausted = true
			break
		}
		items = append(items, value)
	}
	return map[string]any{"items": items, "fetched_pages": iterator.fetched, "exhausted": exhausted}, nil
}
